File: crawler/security/url_normalizer.py
"""Pure, network-free outbound URL normalization for TASK-022B."""

from __future__ import annotations

import ipaddress
import re
import string
import unicodedata

import idna

from crawler.security.models import NormalizedURL

MAX_URL_UTF8_BYTES = 8192
MAX_HOSTNAME_OCTETS = 253
MAX_LABEL_OCTETS = 63

_UNRESERVED = set(string.ascii_letters + string.digits + "-._~")
_HEX = set("0123456789abcdefABCDEF")
_NUMERIC_DOTTED = re.compile(r"^[0-9.]+$")
_HEX_IP = re.compile(r"^0[xX][0-9a-fA-F]+$")
_MIXED_HEX_IP = re.compile(r"^[0-9a-fA-F.xX]+$")
_IPV4_QUAD = re.compile(r"^([0-9]{1,3})\.([0-9]{1,3})\.([0-9]{1,3})\.([0-9]{1,3})$")
_SCHEME_RE = re.compile(r"^([A-Za-z][A-Za-z0-9+.\-]*):")


class URLNormalizationError(ValueError):
    def __init__(self, reason_code: str, message: str = "") -> None:
        super().__init__(message or reason_code)
        self.reason_code = reason_code


def _fail(reason_code: str) -> None:
    raise URLNormalizationError(reason_code)


def _reject_control_and_backslash(raw: str) -> None:
    for ch in raw:
        cat = unicodedata.category(ch)
        if cat in ("Cc", "Cf", "Zs", "Zl", "Zp") or ch.isspace():
            _fail("control_character")
        if ch == "\\":
            _fail("backslash_not_allowed")


def _parse_strict_ipv4(host: str) -> str | None:
    match = _IPV4_QUAD.fullmatch(host)
    if not match:
        return None
    parts = list(match.groups())
    if any(len(part) > 1 and part.startswith("0") for part in parts):
        return None
    values = [int(part) for part in parts]
    if any(value > 255 for value in values):
        return None
    return ".".join(str(value) for value in values)


def _looks_numeric_ambiguous(host: str) -> bool:
    if _NUMERIC_DOTTED.fullmatch(host):
        return True
    if _HEX_IP.fullmatch(host):
        return True
    if _MIXED_HEX_IP.fullmatch(host) and ("x" in host.lower()):
        return True
    return False


_UNICODE_LABEL_SEPARATORS = {".", "\u3002", "\uff0e", "\uff61"}


def _looks_like_unicode_numeric_host(value: str) -> bool:
    has_non_ascii_digit = False
    remaining = value[:-1] if value.endswith(".") else value
    for ch in remaining:
        if ch.isdecimal() and not ch.isascii():
            has_non_ascii_digit = True
        if ch.isdecimal() or ch in _UNICODE_LABEL_SEPARATORS:
            continue
        return False
    return has_non_ascii_digit and bool(remaining)


def _normalize_arabic_indic_digits(value: str) -> str:
    out: list[str] = []
    for ch in value:
        code = ord(ch)
        if 0x0660 <= code <= 0x0669:
            out.append(chr(ord("0") + code - 0x0660))
        elif 0x06F0 <= code <= 0x06F9:
            out.append(chr(ord("0") + code - 0x06F0))
        else:
            out.append(ch)
    return "".join(out)


def _parse_authority(authority: str) -> tuple[str, int | None]:
    if not authority:
        _fail("authority_required")
    if "@" in authority:
        _fail("userinfo_not_allowed")

    if authority.startswith("["):
        close = authority.find("]")
        if close < 0:
            _fail("invalid_url")
        host = authority[1:close]
        suffix = authority[close + 1 :]
        port: int | None = None
        if suffix:
            if not suffix.startswith(":") or len(suffix) == 1:
                _fail("invalid_port")
            port = _parse_port(suffix[1:])
        if not host:
            _fail("invalid_host")
        return host, port

    if "[" in authority or "]" in authority:
        _fail("invalid_host")
    colon = authority.rfind(":")
    if colon >= 0:
        if authority.count(":") > 1:
            _fail("invalid_host")
        host = authority[:colon]
        port = _parse_port(authority[colon + 1 :])
    else:
        host = authority
        port = None
    if not host:
        _fail("invalid_host")
    return host, port


def _parse_port(value: str) -> int:
    if not value or not re.fullmatch(r"[0-9]+", value):
        _fail("invalid_port")
    port = int(value)
    if port <= 0 or port > 65535:
        _fail("invalid_port")
    return port


def _parse_raw_url(raw: str) -> tuple[str, str, str, str]:
    match = _SCHEME_RE.match(raw)
    if not match:
        _fail("absolute_url_required")
    scheme = match.group(1).lower()
    if scheme not in ("http", "https"):
        _fail("scheme_not_allowed")
    rest = raw[match.end() :]
    if not rest.startswith("//"):
        _fail("absolute_url_required")
    rest = rest[2:]
    auth_end = len(rest)
    for marker in ("/", "?", "#"):
        pos = rest.find(marker)
        if pos >= 0 and pos < auth_end:
            auth_end = pos
    authority = rest[:auth_end]
    path_query = rest[auth_end:]
    if "#" in path_query:
        path_query = path_query[: path_query.index("#")]
    if "?" in path_query:
        path, query = path_query.split("?", 1)
    else:
        path, query = path_query, ""
    return scheme, authority, path, query


def _normalize_host(raw_authority: str, host: str) -> tuple[str, bool]:
    if not host:
        _fail("invalid_host")
    if "%" in host:
        _fail("invalid_host")

    if raw_authority.startswith("["):
        try:
            addr = ipaddress.IPv6Address(host)
        except ValueError:
            _fail("invalid_host")
        return addr.compressed, True

    strict_ipv4 = _parse_strict_ipv4(host)
    if strict_ipv4 is not None:
        return strict_ipv4, True

    if _looks_like_unicode_numeric_host(host):
        _fail("ambiguous_ip_literal")
    if _looks_numeric_ambiguous(host):
        _fail("ambiguous_ip_literal")

    root_dot = host.endswith(".")
    candidate = host[:-1] if root_dot else host
    if not candidate:
        _fail("invalid_host")
    if candidate.startswith(".") or ".." in candidate:
        _fail("invalid_host")
    if len(candidate.encode("utf-8")) > MAX_HOSTNAME_OCTETS:
        _fail("host_too_long")
    labels = candidate.split(".")
    if not labels or any(not label for label in labels):
        _fail("invalid_host")
    for label in labels:
        if len(label.encode("utf-8")) > MAX_LABEL_OCTETS:
            _fail("label_too_long")
    candidate = _normalize_arabic_indic_digits(candidate)
    try:
        ascii_host = idna.encode(candidate, uts46=True, std3_rules=True, transitional=False).decode("ascii").lower()
    except Exception:
        _fail("invalid_idna")
    if not ascii_host or len(ascii_host.encode("ascii")) > MAX_HOSTNAME_OCTETS:
        _fail("host_too_long")
    for label in ascii_host.split("."):
        if not label or len(label.encode("ascii")) > MAX_LABEL_OCTETS:
            _fail("label_too_long")
        if label.startswith("xn--"):
            try:
                decoded = idna.decode(label, uts46=True, std3_rules=True)
            except Exception:
                _fail("invalid_idna")
            if not any(ord(ch) > 127 for ch in decoded):
                _fail("invalid_idna")
    if _parse_strict_ipv4(ascii_host) is not None or _looks_numeric_ambiguous(ascii_host):
        _fail("ambiguous_ip_literal")
    return ascii_host, False


def _canonical_percent(value: str) -> str:
    out: list[str] = []
    i = 0
    while i < len(value):
        ch = value[i]
        if ch == "%":
            if i + 2 >= len(value) or value[i + 1] not in _HEX or value[i + 2] not in _HEX:
                _fail("invalid_percent_encoding")
            code = int(value[i + 1 : i + 3], 16)
            if code < 0x20 or code == 0x7F:
                _fail("invalid_percent_encoding")
            char = chr(code)
            if char in _UNRESERVED:
                out.append(char)
            else:
                out.append("%" + value[i + 1 : i + 3].upper())
            i += 3
            continue
        if ch in _UNRESERVED:
            out.append(ch)
        elif ord(ch) < 128:
            if unicodedata.category(ch) in ("Cc", "Cf"):
                _fail("invalid_percent_encoding")
            out.append(ch)
        else:
            for byte in ch.encode("utf-8"):
                out.append("%" + format(byte, "02X"))
        i += 1
    return "".join(out)


def _remove_last_segment(output: str) -> str:
    idx = output.rfind("/")
    if idx < 0:
        return ""
    return output[:idx]


def _remove_dot_segments(path: str) -> str:
    remaining = path
    output = ""
    while remaining:
        if remaining.startswith("../"):
            remaining = remaining[3:]
        elif remaining.startswith("./"):
            remaining = remaining[2:]
        elif remaining.startswith("/./"):
            remaining = "/" + remaining[3:]
        elif remaining == "/.":
            remaining = "/"
        elif remaining.startswith("/../"):
            remaining = "/" + remaining[4:]
            output = _remove_last_segment(output)
        elif remaining == "/..":
            remaining = "/"
            output = _remove_last_segment(output)
        elif remaining in (".", ".."):
            remaining = ""
        else:
            if remaining.startswith("/"):
                next_slash = remaining.find("/", 1)
                if next_slash < 0:
                    output += remaining
                    remaining = ""
                else:
                    output += remaining[:next_slash]
                    remaining = remaining[next_slash:]
            else:
                next_slash = remaining.find("/")
                if next_slash < 0:
                    output += remaining
                    remaining = ""
                else:
                    output += remaining[:next_slash]
                    remaining = remaining[next_slash:]
    return output


def normalize_outbound_url(raw: str) -> NormalizedURL:
    if not isinstance(raw, str):
        _fail("invalid_url")
    if raw == "":
        _fail("empty_url")
    if len(raw.encode("utf-8")) > MAX_URL_UTF8_BYTES:
        _fail("url_too_long")
    _reject_control_and_backslash(raw)

    scheme, authority, path, query = _parse_raw_url(raw)
    if not authority:
        _fail("authority_required")
    host, raw_port = _parse_authority(authority)
    port = raw_port if raw_port is not None else (80 if scheme == "http" else 443)
    explicit_port = raw_port is not None

    host, is_ip_literal = _normalize_host(authority, host)
    host_for_authority = f"[{host}]" if is_ip_literal and ":" in host else host
    canonical_authority = host_for_authority
    default_port = 80 if scheme == "http" else 443
    if explicit_port and port != default_port:
        canonical_authority = f"{host_for_authority}:{port}"

    canonical_path = _canonical_percent(path)
    canonical_query = _canonical_percent(query)
    canonical_path = _remove_dot_segments(canonical_path)
    if not canonical_path:
        canonical_path = "/"

    normalized_url = scheme + "://" + canonical_authority + canonical_path
    if canonical_query:
        normalized_url += "?" + canonical_query
    if len(normalized_url.encode("utf-8")) > MAX_URL_UTF8_BYTES:
        _fail("url_too_long")

    return NormalizedURL(
        scheme=scheme,
        host=host,
        port=port,
        explicit_port=explicit_port,
        authority=canonical_authority,
        normalized_url=normalized_url,
        is_ip_literal=is_ip_literal,
    )

File: go-spider/internal/security/url_normalizer.go
package security

import (
	"fmt"
	"net/netip"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"unicode"

	"golang.org/x/net/idna"
)

var schemePattern = regexp.MustCompile(`^([A-Za-z][A-Za-z0-9+.\-]*):`)

const (
	MaxURLUTF8Bytes   = 8192
	MaxHostnameOctets = 253
	MaxLabelOctets    = 63
)

type URLNormalizationError struct {
	ReasonCode string
}

func (e *URLNormalizationError) Error() string { return e.ReasonCode }

func fail(reason string) error { return &URLNormalizationError{ReasonCode: reason} }

func schemeFromRaw(raw string) (string, error) {
	match := schemePattern.FindStringSubmatch(raw)
	if match == nil {
		return "", nil
	}
	return strings.ToLower(match[1]), nil
}

func isUnicodeLabelSeparator(r rune) bool {
	return r == '.' || r == '\u3002' || r == '\uff0e' || r == '\uff61'
}

func looksLikeUnicodeNumericHost(value string) bool {
	if strings.HasSuffix(value, ".") {
		value = value[:len(value)-1]
	}
	if value == "" {
		return false
	}
	hasNonASCII := false
	for _, r := range value {
		if unicode.Is(unicode.Nd, r) {
			if r > 127 {
				hasNonASCII = true
			}
			continue
		}
		if isUnicodeLabelSeparator(r) {
			continue
		}
		return false
	}
	return hasNonASCII
}

func normalizeArabicIndicDigits(value string) string {
	var b strings.Builder
	for _, r := range value {
		switch {
		case r >= 0x0660 && r <= 0x0669:
			b.WriteRune('0' + (r - 0x0660))
		case r >= 0x06F0 && r <= 0x06F9:
			b.WriteRune('0' + (r - 0x06F0))
		default:
			b.WriteRune(r)
		}
	}
	return b.String()
}

func isHexByte(b byte) bool {
	return (b >= '0' && b <= '9') || (b >= 'a' && b <= 'f') || (b >= 'A' && b <= 'F')
}

func hexValue(b byte) int {
	switch {
	case b >= '0' && b <= '9':
		return int(b - '0')
	case b >= 'a' && b <= 'f':
		return int(b-'a') + 10
	default:
		return int(b-'A') + 10
	}
}

func upperHex(b byte) byte {
	if b >= 'a' && b <= 'f' {
		return b - 'a' + 'A'
	}
	return b
}

func isUnreserved(b byte) bool {
	return (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || (b >= '0' && b <= '9') || b == '-' || b == '.' || b == '_' || b == '~'
}

func rejectControlAndBackslash(raw string) error {
	for _, r := range raw {
		if r == '\\' {
			return fail("backslash_not_allowed")
		}
		if unicode.IsSpace(r) || unicode.IsControl(r) || unicode.Is(unicode.Cf, r) {
			return fail("control_character")
		}
	}
	return nil
}

func parseStrictIPv4(host string) (string, bool) {
	parts := strings.Split(host, ".")
	if len(parts) != 4 {
		return "", false
	}
	out := make([]string, 4)
	for i, part := range parts {
		if part == "" || len(part) > 3 {
			return "", false
		}
		if len(part) > 1 && part[0] == '0' {
			return "", false
		}
		n, err := strconv.Atoi(part)
		if err != nil || n < 0 || n > 255 {
			return "", false
		}
		out[i] = strconv.Itoa(n)
	}
	return strings.Join(out, "."), true
}

func looksNumericAmbiguous(host string) bool {
	digitsDots := true
	hasDot := false
	for _, r := range host {
		if r >= '0' && r <= '9' {
			continue
		}
		if r == '.' {
			hasDot = true
			continue
		}
		digitsDots = false
		break
	}
	if digitsDots {
		return true
	}
	if strings.HasPrefix(host, "0x") || strings.HasPrefix(host, "0X") {
		rest := host[2:]
		allHex := rest != ""
		for _, r := range rest {
			if !isHexByte(byte(r)) {
				allHex = false
				break
			}
		}
		if allHex {
			return true
		}
	}
	if strings.Contains(strings.ToLower(host), "x") {
		allMixed := true
		for _, r := range host {
			if isHexByte(byte(r)) || r == '.' || r == 'x' || r == 'X' {
				continue
			}
			allMixed = false
			break
		}
		if allMixed {
			return true
		}
	}
	_ = hasDot
	return false
}

func normalizeHost(rawAuthority, host string) (string, bool, error) {
	if host == "" {
		return "", false, fail("invalid_host")
	}
	if strings.Contains(host, "%") {
		return "", false, fail("invalid_host")
	}

	if strings.HasPrefix(rawAuthority, "[") {
		addr, err := netip.ParseAddr(host)
		if err != nil || !addr.Is6() {
			return "", false, fail("invalid_host")
		}
		return addr.String(), true, nil
	}

	if quad, ok := parseStrictIPv4(host); ok {
		return quad, true, nil
	}
	if looksLikeUnicodeNumericHost(host) {
		return "", false, fail("ambiguous_ip_literal")
	}
	if looksNumericAmbiguous(host) {
		return "", false, fail("ambiguous_ip_literal")
	}

	candidate := host
	if strings.HasSuffix(candidate, ".") {
		candidate = candidate[:len(candidate)-1]
	}
	if candidate == "" || strings.HasPrefix(candidate, ".") || strings.Contains(candidate, "..") {
		return "", false, fail("invalid_host")
	}
	if len(candidate) > MaxHostnameOctets {
		return "", false, fail("host_too_long")
	}
	labels := strings.Split(candidate, ".")
	for _, label := range labels {
		if label == "" {
			return "", false, fail("invalid_host")
		}
		if len(label) > MaxLabelOctets {
			return "", false, fail("label_too_long")
		}
	}

	candidate = normalizeArabicIndicDigits(candidate)
	asciiHost, err := idna.Lookup.ToASCII(candidate)
	if err != nil {
		return "", false, fail("invalid_idna")
	}
	asciiHost = strings.ToLower(asciiHost)
	if asciiHost == "" || len(asciiHost) > MaxHostnameOctets {
		return "", false, fail("host_too_long")
	}
	for _, label := range strings.Split(asciiHost, ".") {
		if label == "" {
			return "", false, fail("invalid_host")
		}
		if len(label) > MaxLabelOctets {
			return "", false, fail("label_too_long")
		}
		if strings.HasPrefix(label, "xn--") {
			decoded, err := idna.Lookup.ToUnicode(label)
			if err != nil || !containsNonASCII(decoded) {
				return "", false, fail("invalid_idna")
			}
		}
	}
	if _, ok := parseStrictIPv4(asciiHost); ok || looksNumericAmbiguous(asciiHost) {
		return "", false, fail("ambiguous_ip_literal")
	}
	return asciiHost, false, nil
}

func containsNonASCII(s string) bool {
	for _, r := range s {
		if r > 127 {
			return true
		}
	}
	return false
}

func canonicalPercent(value string) (string, error) {
	var b strings.Builder
	for i := 0; i < len(value); {
		c := value[i]
		if c == '%' {
			if i+2 >= len(value) || !isHexByte(value[i+1]) || !isHexByte(value[i+2]) {
				return "", fail("invalid_percent_encoding")
			}
			code := hexValue(value[i+1])<<4 | hexValue(value[i+2])
			if code < 0x20 || code == 0x7F {
				return "", fail("invalid_percent_encoding")
			}
			if isUnreserved(byte(code)) {
				b.WriteByte(byte(code))
			} else {
				b.WriteByte('%')
				b.WriteByte(upperHex(value[i+1]))
				b.WriteByte(upperHex(value[i+2]))
			}
			i += 3
			continue
		}
		if c < 0x80 {
			if c < 0x20 || c == 0x7F {
				return "", fail("invalid_percent_encoding")
			}
			b.WriteByte(c)
			i++
			continue
		}
		fmt.Fprintf(&b, "%%%02X", c)
		i++
	}
	return b.String(), nil
}

func removeLastSegment(output string) string {
	idx := strings.LastIndex(output, "/")
	if idx < 0 {
		return ""
	}
	return output[:idx]
}

func removeDotSegments(path string) string {
	remaining := path
	output := ""
	for remaining != "" {
		switch {
		case strings.HasPrefix(remaining, "../"):
			remaining = remaining[3:]
		case strings.HasPrefix(remaining, "./"):
			remaining = remaining[2:]
		case strings.HasPrefix(remaining, "/./"):
			remaining = "/" + remaining[3:]
		case remaining == "/.":
			remaining = "/"
		case strings.HasPrefix(remaining, "/../"):
			remaining = "/" + remaining[4:]
			output = removeLastSegment(output)
		case remaining == "/..":
			remaining = "/"
			output = removeLastSegment(output)
		case remaining == "." || remaining == "..":
			remaining = ""
		default:
			if strings.HasPrefix(remaining, "/") {
				next := strings.Index(remaining[1:], "/")
				if next < 0 {
					output += remaining
					remaining = ""
				} else {
					idx := next + 1
					output += remaining[:idx]
					remaining = remaining[idx:]
				}
			} else {
				next := strings.Index(remaining, "/")
				if next < 0 {
					output += remaining
					remaining = ""
				} else {
					output += remaining[:next]
					remaining = remaining[next:]
				}
			}
		}
	}
	return output
}

func splitRawPathQuery(raw string) (string, string) {
	idx := strings.Index(raw, "://")
	if idx < 0 {
		return "", ""
	}
	rest := raw[idx+3:]
	authEnd := strings.IndexAny(rest, "/?#")
	if authEnd < 0 {
		return "", ""
	}
	rest = rest[authEnd:]
	if i := strings.IndexByte(rest, '#'); i >= 0 {
		rest = rest[:i]
	}
	if i := strings.IndexByte(rest, '?'); i >= 0 {
		return rest[:i], rest[i+1:]
	}
	return rest, ""
}

func authorityFromRaw(raw string) string {
	idx := strings.Index(raw, "://")
	if idx < 0 {
		return ""
	}
	authority := raw[idx+3:]
	if end := strings.IndexAny(authority, "/?#"); end >= 0 {
		authority = authority[:end]
	}
	return authority
}

func pathQueryHasInvalidPercent(raw string) bool {
	path, query := splitRawPathQuery(raw)
	if _, err := canonicalPercent(path); err != nil {
		return true
	}
	if _, err := canonicalPercent(query); err != nil {
		return true
	}
	return false
}

func validateRawPort(raw string) error {
	idx := strings.Index(raw, "://")
	if idx < 0 {
		return nil
	}
	authority := raw[idx+3:]
	if end := strings.IndexAny(authority, "/?#"); end >= 0 {
		authority = authority[:end]
	}
	if strings.Contains(authority, "@") {
		return nil
	}
	if strings.HasPrefix(authority, "[") {
		closeBracket := strings.Index(authority, "]")
		if closeBracket < 0 {
			return fail("invalid_url")
		}
		suffix := authority[closeBracket+1:]
		if suffix == "" {
			return nil
		}
		if !strings.HasPrefix(suffix, ":") {
			return fail("invalid_port")
		}
		portStr := suffix[1:]
		if !validPortString(portStr) {
			return fail("invalid_port")
		}
		return nil
	}
	colon := strings.LastIndex(authority, ":")
	if colon < 0 {
		return nil
	}
	portStr := authority[colon+1:]
	if !validPortString(portStr) {
		return fail("invalid_port")
	}
	return nil
}

func validPortString(portStr string) bool {
	if portStr == "" {
		return false
	}
	for i := 0; i < len(portStr); i++ {
		if portStr[i] < '0' || portStr[i] > '9' {
			return false
		}
	}
	n, err := strconv.Atoi(portStr)
	return err == nil && n > 0 && n <= 65535
}

func NormalizeOutboundURL(raw string) (NormalizedURL, error) {
	if raw == "" {
		return NormalizedURL{}, fail("empty_url")
	}
	if len([]byte(raw)) > MaxURLUTF8Bytes {
		return NormalizedURL{}, fail("url_too_long")
	}
	if err := rejectControlAndBackslash(raw); err != nil {
		return NormalizedURL{}, err
	}
	scheme, schemeErr := schemeFromRaw(raw)
	if schemeErr != nil {
		return NormalizedURL{}, schemeErr
	}
	if scheme == "" {
		return NormalizedURL{}, fail("absolute_url_required")
	}
	if scheme != "http" && scheme != "https" {
		return NormalizedURL{}, fail("scheme_not_allowed")
	}
	if err := validateRawPort(raw); err != nil {
		return NormalizedURL{}, err
	}
	u, err := url.Parse(raw)
	if err != nil {
		if pathQueryHasInvalidPercent(raw) {
			return NormalizedURL{}, fail("invalid_percent_encoding")
		}
		if strings.Contains(authorityFromRaw(raw), "%") {
			return NormalizedURL{}, fail("invalid_host")
		}
		return NormalizedURL{}, fail("invalid_url")
	}
	scheme = strings.ToLower(u.Scheme)
	if !strings.Contains(raw, "://") {
		return NormalizedURL{}, fail("absolute_url_required")
	}
	if u.Host == "" {
		return NormalizedURL{}, fail("authority_required")
	}
	if u.User != nil || strings.Contains(u.Host, "@") {
		return NormalizedURL{}, fail("userinfo_not_allowed")
	}
	hostname := u.Hostname()
	if hostname == "" {
		return NormalizedURL{}, fail("invalid_host")
	}

	if strings.HasSuffix(u.Host, ":") {
		return NormalizedURL{}, fail("invalid_port")
	}
	rawPort := u.Port()
	var port int
	explicitPort := rawPort != ""
	if explicitPort {
		parsed, err := strconv.Atoi(rawPort)
		if err != nil || parsed <= 0 || parsed > 65535 {
			return NormalizedURL{}, fail("invalid_port")
		}
		port = parsed
	} else if scheme == "http" {
		port = 80
	} else {
		port = 443
	}

	host, isIPLiteral, err := normalizeHost(u.Host, hostname)
	if err != nil {
		return NormalizedURL{}, err
	}
	hostForAuthority := host
	if isIPLiteral && strings.Contains(host, ":") {
		hostForAuthority = "[" + host + "]"
	}
	authority := hostForAuthority
	defaultPort := 80
	if scheme == "https" {
		defaultPort = 443
	}
	if explicitPort && port != defaultPort {
		authority = fmt.Sprintf("%s:%d", hostForAuthority, port)
	}

	path, query := splitRawPathQuery(raw)
	canonicalPath, err := canonicalPercent(path)
	if err != nil {
		return NormalizedURL{}, err
	}
	canonicalQuery, err := canonicalPercent(query)
	if err != nil {
		return NormalizedURL{}, err
	}
	canonicalPath = removeDotSegments(canonicalPath)
	if canonicalPath == "" || canonicalPath == "." {
		canonicalPath = "/"
	}

	normalized := scheme + "://" + authority + canonicalPath
	if canonicalQuery != "" {
		normalized += "?" + canonicalQuery
	}
	if len([]byte(normalized)) > MaxURLUTF8Bytes {
		return NormalizedURL{}, fail("url_too_long")
	}

	return NormalizedURL{
		Scheme:        scheme,
		Host:          host,
		Port:          port,
		ExplicitPort:  explicitPort,
		Authority:     authority,
		NormalizedURL: normalized,
		IsIPLiteral:   isIPLiteral,
	}, nil
}
